### PR TITLE
Fix stripe ICharge interface with missing fields and optional null values

### PR DESCRIPTION
**Pull request by galtalmor**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [ ] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [ ] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [ ] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [ ] `tslint.json` should be present, and `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://stripe.com/docs/api#charges
- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

If removing a declaration:
- [ ] If a package was never on DefinitelyTyped, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [ ] Delete the package's directory.
- [ ] Add it to `notNeededPackages.json`.

### CODE REVIEW

File: types/stripe/index.d.ts [hunk 1]
```diff
@@ -1056,7 +1097,7 @@ declare namespace Stripe {
             }
         }
 
-        interface IChargeRefunds extends IList<refunds.IRefund>, resources.ChargeRefunds { }
```
Comment (CodeAnimal) [inline types/stripe/index.d.ts]: Why did you remove the `resources.ChargeRefunds` inheritance?

File: types/stripe/index.d.ts [hunk 2]
```diff
@@ -842,10 +870,15 @@ declare namespace Stripe {
              */
             refunds: IChargeRefunds;
 
+            /**
+             * ID of the review associated with this charge if one exists. [Expandable]
+             */
+            review?: string | null;
```
Comment (CodeAnimal) [inline types/stripe/index.d.ts]: This property is `Expandable` which means it can be expanded into an object (see `source_transfer` on L894 as an example). Can you also add the object interface to the type (creating the interface if need be)?

File: types/stripe/index.d.ts [hunk 3]
```diff
@@ -767,25 +773,33 @@ declare namespace Stripe {
...
 
             description?: string;
 
+            /**
+             * The account (if any) the charge was made on behalf of, with an automatic
+             * transfer. See the [Connect documentation]
+             * <https://stripe.com/docs/connect/destination-charges> for details.
+             * [Expandable]
+             */
+            destination?: string | null;
```
Comment (CodeAnimal) [inline types/stripe/index.d.ts]: This property is `Expandable` which means it can be expanded into an object (see `source_transfer` on L894 as an example). Can you also add the object interface to the type (creating the interface if need be)?

File: types/stripe/index.d.ts [hunk 4]
```diff
@@ -737,12 +738,17 @@ declare namespace Stripe {
              * charge if a partial refund was issued), positive integer or zero.
              */
             amount_refunded: number;
+            
+            /**
+             * ID of the Connect application that created the charge. [Expandable]
+             */
+            application?: string | null;
```
Comment (CodeAnimal) [inline types/stripe/index.d.ts]: This property is `Expandable` which means it can be expanded into an object (see `source_transfer` on L894 as an example). Can you also add the object interface to the type (creating the interface if need be)?

File: types/stripe/index.d.ts [hunk 5]
```diff
@@ -805,16 +819,30 @@ declare namespace Stripe {
...
              * ID of the order this charge is for if one exists. [Expandable]
              */
-            order: string | orders.IOrder;
+            order: string | orders.IOrder | null;
+
+            /**
+             * Details about whether the payment was accepted, and why. See
+             * understanding declines for details. [Expandable]
+             */
+            outcome?: any;
```
Comment (CodeAnimal) [inline types/stripe/index.d.ts]: This outcome is actually a hash object with required and optional fields (https://stripe.com/docs/api#charge_object-outcome). This should be made into an interface and set that as the type for the `outcome` field.

**Review by CodeAnimal — CHANGES_REQUESTED** [review]
Hi, thanks for the PR!

Have a look at the comments on individual lines. I'd say they were necessary for this PR to be accepted, along with updates to the tests file. This is to ensure that the error you're receiving is reproducible in the tests, and therefore corrected by your changes. This will also ensure the error doesn't creep back into the definitions file. The tests would need to cover the changes I've suggested too. Without updates to the tests file this PR couldn't be accept.

**Comment by galtalmor** [comment]
@CodeAnimal Thanks for the comments. All fixed.
Cheers.

File: types/stripe/index.d.ts [hunk 6]
```diff
@@ -4979,6 +5021,59 @@ declare namespace Stripe {
...
+
+            /**
+             * Time at which the object was created. Measured in seconds since the Unix epoch.
+             */
+            "created": number;
+
+            /**
+             * Has the value true if the object exists in live mode or the value false if the object exists in test mode.
+             */
+            "livemode": false;
```
Comment (LinusU) [inline types/stripe/index.d.ts]: This should probably be `boolean`?

File: types/stripe/index.d.ts [hunk 7]
```diff
@@ -4979,6 +5021,59 @@ declare namespace Stripe {
...
+
+            /**
+             * If true, the review needs action.
+             */
+            "open": true;
```
Comment (LinusU) [inline types/stripe/index.d.ts]: Here too?

File: types/stripe/index.d.ts [hunk 8]
```diff
@@ -4979,6 +5021,59 @@ declare namespace Stripe {
         }
     }
 
+    namespace reviews {
+        interface IReview extends IResourceObject {
+            /**
+             * Unique identifier for the object.
+             */
+            "id": string;
```
Comment (CodeAnimal) [inline types/stripe/index.d.ts]: This is strictly not necessary as it is already present from the IResourceObject interface.

File: types/stripe/index.d.ts [hunk 9]
```diff
@@ -4979,6 +5021,59 @@ declare namespace Stripe {
...
+
+            /**
+             * The reason the review is currently open or closed.
+             */
+            "reason": "rule" | "manual" | "approved" | "refunded" | "refunded_as_fraud" | "disputed";
+        }
+    }
+
+    namespace applications {
+        interface IApplication extends IResourceObject {
```
Comment (CodeAnimal) [inline types/stripe/index.d.ts]: Have you got any documentation for where you got these properties? I can't seem to find any application object documented on the Stripe API docs.

Comment (CodeAnimal) [reply types/stripe/index.d.ts]: I see why you've done this though, as the `application` property is `Expandable`

File: types/stripe/index.d.ts [hunk 10]
```diff
@@ -4979,6 +5021,59 @@ declare namespace Stripe {
         }
     }
 
+    namespace reviews {
+        interface IReview extends IResourceObject {
```
Comment (CodeAnimal) [inline types/stripe/index.d.ts]: The field names don't need quotes around them.

File: types/stripe/index.d.ts [hunk 11]
```diff
@@ -5382,18 +5518,18 @@ declare namespace Stripe {
              * Once entirely refunded, a charge can't be refunded again.
              * This method will throw an error when called on an already-refunded charge, or when trying to refund more money than is left on a charge.
              */
-            create(data: refunds.IRefundCreationOptions, options: HeaderOptions, response?: IResponseFn<refunds.IRefund>): Promise<refunds.IRefund>;
-            create(data: refunds.IRefundCreationOptions, response?: IResponseFn<refunds.IRefund>): Promise<refunds.IRefund>;
-            create(options: HeaderOptions, response?: IResponseFn<refunds.IRefund>): Promise<refunds.IRefund>;
-            create(response?: IResponseFn<refunds.IRefund>): Promise<refunds.IRefund>;
+            create?(data: refunds.IRefundCreationOptions, options: HeaderOptions, response?: IResponseFn<refunds.IRefund>): Promise<refunds.IRefund>;
```
Comment (CodeAnimal) [inline types/stripe/index.d.ts]: Why have these been made optional?

**Review by CodeAnimal — CHANGES_REQUESTED** [review]
Hey thanks for the updates!

I'm still not sure why the null types are being added to the definitions. Can you explain why?

You still need to add the tests to ensure that the changes stay. The tests need to fail without the changes, but pass with them - therefore demonstrating the situation why the null definitions are needed in the first place and also makes sure other additions are removed in the future. Without tests this PR shouldn't be accepted.

Take a look at the other line comments too :)

File: types/stripe/index.d.ts (hunk 6 above)
Comment (CodeAnimal) [reply types/stripe/index.d.ts]: I agree

File: types/stripe/index.d.ts (hunk 7 above)
Comment (CodeAnimal) [reply types/stripe/index.d.ts]: 👍

File: types/stripe/index.d.ts (hunk 9 above)
Comment (galtalmor) [reply types/stripe/index.d.ts]: Yes. Actually I figured this out from the .Net API docs.

File: types/stripe/index.d.ts (hunk 11 above)
Comment (galtalmor) [reply types/stripe/index.d.ts]: Well, this is primarily a data model object. For some reason when I actually use the library I've noticed the quite often I don't really have these functions available. This is why I did that.

**Comment by galtalmor** [comment]
@CodeAnimal @LinusU Thank you for the comments.
Did my best to fix/reply to all.
Cheers.
